packages/api/src/services/subscriptions.ts

```diff
@@ -96,42 +98,63 @@ export const saveSubscription = async ({
   unsubscribeHttpUrl,
   icon,
 }: SaveSubscriptionInput): Promise<string> => {
-  const result = await getRepository(Subscription).upsert(
-    {
-      name,
-      newsletterEmail: { id: newsletterEmail.id },
-      user: { id: userId },
-      unsubscribeHttpUrl,
-      unsubscribeMailTo,
-      icon,
-    },
-    ['name', 'user']
+  const subscriptionData = {
+    unsubscribeHttpUrl,
+    unsubscribeMailTo,
+    icon,
+    lastFetchedAt: new Date(),
+  }
+
+  const existingSubscription = await getSubscriptionByNameAndUserId(
+    name,
+    userId
   )
+  if (existingSubscription) {
+    // update subscription if already exists
+    await getRepository(Subscription).update(
+      existingSubscription.id,
+      subscriptionData
+    )
+
+    return existingSubscription.id
+  }
+
+  const result = await getRepository(Subscription).save({
+    ...subscriptionData,
+    name,
+    newsletterEmail: { id: newsletterEmail.id },
+    user: { id: userId },
+    type: SubscriptionType.Newsletter,
+  })
 
-  return result.identifiers[0].id as string
+  return result.id
 }
 
 export const unsubscribe = async (subscription: Subscription) => {
-  let unsubscribed = false
-  if (subscription.unsubscribeMailTo) {
-    // unsubscribe by sending email
-    unsubscribed = await sendUnsubscribeEmail(
-      subscription.unsubscribeMailTo,
-      subscription.newsletterEmail.address
-    )
-  }
-  // TODO: find a good way to unsubscribe by url if email fails or not provided
-  // because it often requires clicking a button on the page to unsubscribe
-
-  if (!unsubscribed) {
-    // update subscription status to unsubscribed if failed to unsubscribe
-    console.log('Failed to unsubscribe', subscription.id)
-    return getRepository(Subscription).update(subscription.id, {
-      status: SubscriptionStatus.Unsubscribed,
-    })
+  // unsubscribe from newsletter
+  if (subscription.type === SubscriptionType.Newsletter) {
+    let unsubscribed = false
+
+    if (subscription.unsubscribeMailTo && subscription.newsletterEmail) {
+      // unsubscribe by sending email
+      unsubscribed = await sendUnsubscribeEmail(
+        subscription.unsubscribeMailTo,
+        subscription.newsletterEmail.address
+      )
+    }
+    // TODO: find a good way to unsubscribe by url if email fails or not provided
```

Comment:
I think in this case we might need to return the URL to the client, that's at least what I have seen GMail do

Reply:
yeah, and we can open the URL on the client side